lib/kennel/syncer.rb

```diff
@@ -130,45 +130,56 @@ def calculate_changes
         populate_id_map @expected, @actual
         filter_actual! @actual
         resolve_linked_tracking_ids! @expected # resolve dependencies to avoid diff
+        @expected.each(&:add_tracking_id) # avoid diff with actual, which has tracking_id
 
-        @expected.each(&:add_tracking_id) # avoid diff with actual
+        # see which expected match the actual
+        matching, unmatched_expected, unmatched_actual = partition_matched_expected
+        validate_expected_id_not_missing unmatched_expected
+        fill_details! matching # need details to diff later
 
-        lookup_map = matching_expected_lookup_map
-        items = @actual.map do |a|
-          e = matching_expected(a, lookup_map)
-          if e && @expected.delete?(e)
-            [e, a]
-          else
-            [nil, a]
-          end
+        # update matching if needed
+        matching.each do |e, a|
+          id = a.fetch(:id)
+          diff = e.diff(a)
+          @update << [id, e, a, diff] if diff.any?
         end
 
-        # fill details of things we need to compare
-        details = items.map { |e, a| a if e && e.class.api_resource == "dashboard" }.compact
-        @api.fill_details! "dashboard", details
+        # delete previously managed
+        unmatched_actual.each { |a| @delete << [a.fetch(:id), nil, a] if a.fetch(:tracking_id) }
 
-        # pick out things to update or delete
-        items.each do |e, a|
-          id = a.fetch(:id)
-          if e
-            diff = e.diff(a) # slow ...
-            if diff.any?
-              @update << [id, e, a, diff]
-            end
-          elsif a.fetch(:tracking_id) # was previously managed
-            @delete << [id, nil, a]
-          end
-        end
+        # unmatched expected need to be created
+        @create = unmatched_expected.map { |e| [nil, e] }
 
-        ensure_all_ids_found
-        @create = @expected.map { |e| [nil, e] }
+        # order to avoid deadlocks
         @delete.sort_by! { |_, _, a| DELETE_ORDER.index a.fetch(:klass).api_resource }
         @update.sort_by! { |_, e, _| DELETE_ORDER.index e.class.api_resource } # slo needs to come before slo alert
       end
     end
 
-    def ensure_all_ids_found
-      @expected.each do |e|
+    def partition_matched_expected
+      lookup_map = matching_expected_lookup_map
+      unmatched_expected = @expected.dup
+      unmatched_actual = []
+      matched = []
+      @actual.each do |a|
+        e = matching_expected(a, lookup_map)
+        if e && unmatched_expected.delete?(e)
+          matched << [e, a]
+        else
+          unmatched_actual << a
+        end
+      end.compact
+      [matched, unmatched_expected, unmatched_actual]
```

Comment:
I like this function :-) 

Mild preference for returning a struct so one can't accidentally get the ordering of the triple wrong.

e.g. `Struct.new(:matched, :unmatched_expected, :unmatched_actual).new(matched, unmatched_expected, unmatched_actual`

It's more verbose, but safer.

Reply:
just looks so ugly 😭

Reply:
all the code is "expected, actual" so doubt the ordering will go wrong 🤞